### PR TITLE
feat: implement `group` component

### LINKED ISSUE #767: Implement `Group` component (also referred to as subform/fieldset)
** Implementation epic: https://github.com/camunda/product-hub/issues/1443 **

### Is your feature request related to a problem? Please describe.

There are multiple topics we want to cover with an implementation of such a component.
- A group would assist with the visual and logical separation of components in the form field. 
- Such a component assists with applying rendering conditions at scale, without having to define them for every child field
- Such a component would assist in the mapping (in and out) of data objects, defining a mapping path in a centralized place.

**Prototype:**
grafik

#### Requirements:
- [x] I can optionally define a path on a group which will map children of the group to the appropriate data
- [x] I can compose nested paths (user.name)
	- [x] For groups
	- [x] For fields
- [x] I can set a hide-if condition on a group, this:
	- [x] Hide the field when the condition is valid 
	- [x] Does not publish the data associated with those fields, following the behavior of the previous fields
- [x] ~Layouting works well with groups~ (removed from scope)
	- [x] I can still use layouting withing groups
	- [x] ~Children of groups still follow the sizing restrictions of the full form, this means that if a group is 8 wide, children can only sum up to 8 wide as well~
	- [x] ~Resizing of the main group should collapse children down to their "minimum" size at most (need to provide a preview of the feel for this)~
- [x] I cannot set fields and paths which conflict
	- [x] Through the properties panel
	- [x] Through dragging
- [x] Fields should remain aligned in and outside of first order groups
- [x] UX wire-frames must be matched

#### Prototype:
Deployed here: https://github.com/bpmn-io/form-js/pull/768

#### Implement:
- [x] Features 
	- [x] Support properties panel configuration of groups
		- [x] Binding path
		    - [x] Value
			- [x] Validate conflicts with other paths / keys
		- [x] Conditional rendering
		- [x] Show outline of group
		- [x] Label
	- [x] Group component to the components list
	- [x] Group component rendering
		- [x] Outlines as part of the editor selection 
		- [x] Outlines as part of rendering
		- [x] First order groups should keep child fields aligned with root
		- [x] Low fidelity implementation
		- [x] Design review of low fidelity implementation
		- [x] Carbonization layer
			- [x] Evaluate the current look (aimed at carbon in the first place)
			- [x] If necessary, add carbonization layer
		- [x] Full design specs + changes implemented [here](https://github.com/camunda/product-hub/issues/1443)
	- [x] Drag and drop
		- [x] Check for path collision
		- [x] Ensure it's intuitive with groups
	- [x] ~Flexible layout~ (removed from scope)
		- [x] ~Restrict sub-grids to that of upper grids~
		- [x] ~Try out multiple resizing behaviors~
			- [x] ~Blocking~
			- [x] ~Collapsing~
	- [x] Variable mapping
		- [x] Paths
			- [x] Empty paths
			- [x] Simple paths (like keys)
			- [x] Composed paths (. operator)
		- [x] Keys
			- [x] Composed keys
		- [x] Data filtering based on hidden fields
		- [x] Ensure `getSchemaVariables` still works as it should
- [x] Bugs
	- [x] Group header is wrong
	- [x] Dragging out of a group causes a crash
	- [x] Dragging into empty path group doesn't work (allow self-conflicts) 
	- [x] Dragging in general is a little janky
	- [x] Groups eat into eachother when placed in the same row (css fixable)
	- [x] Nested groups create empty object variables even when no field is within them
	- [x] It is possible to use array indexes in keys/paths, we should restrict that, or support it fully (which is quite challenging)
	- [x] Can't delete button due to key?
	- [x] Self-conflicting is back
	- [x] From smoke testing
		- [x] Three nested groups with paths a > empty > c with checkbox d inside, hide if won't trigger on a.c.d
		- [x] Resizing groups isn't working properly on chrome / edge, but fine on firefox (alignment issue maybe)? 	
- [ ] Testing
	- [x] Smoke testing via QA 
	- [x] Unit tests
	- [x] UI regression tests
		- [x] basic
		- [x] carbon
	- [ ] Integration testing
- [x] [Linting](https://github.com/camunda/form-linting/pull/73/commits)
	- [x] Disallow group component prior to 1.3
	- [x] Disallow path accessors in keys prior to 1.3
- [x] [Documentation](https://github.com/camunda/camunda-platform-docs/pull/2504)
	- [x] Document group component
	- [x] Document path inheritance
- [ ] Integrate
	- [ ] Release group feature in form-js
	- [x] Release new linting rules
	- [ ] Integrate where needed
	- [ ] smoke test integrations
- [ ] Finish
	- [ ] Raise all follow-ups as tickets
	- [ ] Publish documentation

### Questions
- [x] Should we sacrifice row dragging in favour of more similarity between editor and preview?  - No, we can investigate this further later.

### Follow-ups
- use feel rendering for paths and keys
- only maintain one global grid for sizing
- Disallow self-referential hide-ifs (linting)

### Additional context

- A11y without using a fieldset proper as an underlyling component ([here](https://www.w3.org/WAI/tutorials/forms/grouping/))
- https://github.com/bpmn-io/form-js/issues/252
- https://github.com/bpmn-io/form-js/issues/464

### PR DESCRIPTION
**Pull request by Skaiir**
Closes #767

### CODE REVIEW

**Review by vsgoulart — COMMENTED** [review]
@Skaiir There's the CI issue I mentioned through Slack +2 other issues I found while testing:


1. Fields reduced inside the group block the empty space inside the block

https://github.com/bpmn-io/form-js/assets/1022916/8903955e-b2ec-4dea-93ae-3bfed0428c7e

2. Prefilled nested values are not loaded when blocks are enabled via FEEL condition

https://github.com/bpmn-io/form-js/assets/1022916/b168d34a-09cf-4b27-aa9b-68dc0628ce5e

File: packages/form-js-editor/src/features/properties-panel/entries/KeyEntry.js [hunk 1]
```diff
@@ -70,7 +70,12 @@ function Key(props) {
     }
 
     if (value && !/^\w+(.\w+)*$/.test(value)) {
```
[the remark this answers is not in the document]
Comment (Skaiir) [reply packages/form-js-editor/src/features/properties-panel/entries/KeyEntry.js]: Agreed.

File: packages/form-js-viewer/src/core/Importer.js [hunk 2]
```diff
@@ -32,27 +31,29 @@ export default class Importer {
    * @returns {ImportResult}
    */
   importSchema(schema) {
-
-    // TODO: Add warnings
```
[the remark this answers is not in the document]
Comment (Skaiir) [reply packages/form-js-viewer/src/core/Importer.js:36]: Fair enough, we still want those eventually actually. Will restore, just something that got crunched by my refactoring frenzy.

File: packages/form-js-viewer/src/render/context/FormRenderContext.js [hunk 3]
```diff
@@ -1,6 +1,9 @@
 import { createContext } from 'preact';
 
 const FormRenderContext = createContext({
+  EmptyRoot: (props) => {
```
[the remark this answers is not in the document]
Comment (Skaiir) [reply packages/form-js-viewer/src/render/context/FormRenderContext.js:4]: So we used to have Empty, which was basically rendered inside all grids (which used to be only the root) when they are empty. This is useful for stuff like rendering the "drag in here" image. I split it into two different component, Empty and EmptyRoot, so that we don't render this inside every grid but just the Root one.

grafik

I've kept the Root as we might want to keep this open for external or internal customization in the future.

File: packages/form-js-viewer/test/spec/util/GetSchemaVariables.spec.js [hunk 4]
```diff
@@ -39,7 +40,7 @@ describe('util/getSchemaVariables', () => {
 
     const variables = getSchemaVariables(dynamicSchema);
 
-    expect(variables).to.eql([ 'product', 'xyzData', 'mailto', 'language', 'tags' ]);
+    expect(variables).to.eql([ 'xyzData', 'product', 'mailto', 'language', 'tags' ]);
```
[the remark this answers is not in the document]
Comment (Skaiir) [reply packages/form-js-viewer/test/spec/util/GetSchemaVariables.spec.js:43]: So getSchemaVariables used to linearly iterate through all the components, then search for the keys, and other variables on other properties. This however does not work now because something that's nested within a group isn't a top level key anymore. 

Like group(path: a) > checkbox(key: c), should not export c as a variable as it's a property of root object a. 

So now we iterate through the paths / keys as a separate step from the iteration through all the other properties, which changes the order in which they are aggregated to the array.

File: packages/form-js-carbon-styles/src/carbon-styles.js [hunk 5]
```diff
@@ -1156,14 +1162,6 @@ const CARBON_STYLES = css`
...
     }
 
-    .fjs-layout-column:first-of-type .fjs-form-field {
-      margin-left: 0;
-    }
-
-    .fjs-layout-column:last-of-type .fjs-form-field {
-      margin-right: 0;
-    }
-
```
[the remark this answers is not in the document]
Comment (Skaiir) [reply packages/form-js-carbon-styles/src/carbon-styles.js:1166]: I think this is something that needs to be adjusted in tasklist directly, because what this does is basically stick the fields to the left and right boundaries of the form, which simply cannot work if groups have negative margins.

File: packages/form-js-editor/src/features/modeling/behavior/PathBehavior.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import CommandInterceptor from 'diagram-js/lib/command/CommandInterceptor';
+
+export default class PathBehavior extends CommandInterceptor {
+  constructor(eventBus, modeling) {
+    super(eventBus);
+
+    // @ts-ignore-next-line
```
[the remark this answers is not in the document]
Comment (Skaiir) [reply packages/form-js-editor/src/features/modeling/behavior/PathBehavior.js:7]: Just copied it over to be honest, but it seems something with ts has changed and it's no longer required. I'll remove it from all the behaviors.

File: packages/form-js-viewer/src/core/Importer.js [hunk 7]
```diff
@@ -32,27 +31,29 @@ export default class Importer {
...
     const warnings = [];
 
     try {
-
+      this._cleanup();
```
[the remark this answers is not in the document]
Comment (Skaiir) [reply packages/form-js-viewer/src/core/Importer.js:39]: This has always been necessary and basically a long standing bug that went unnoticed, but basically it made importing things twice impossible as the various registries would not cleanup. Now it's clean and the intent is obvious.

File: package.json [hunk 8]
```diff
@@ -28,6 +28,7 @@
     "start": "npm run start:pkg -- @bpmn-io/form-js",
     "start:viewer": "npm run start:pkg -- @bpmn-io/form-js-viewer",
     "start:viewer:stress": "cross-env SINGLE_START=stress npm run dev -- @bpmn-io/form-js-viewer",
+    "start:viewer:groups": "cross-env SINGLE_START=groups npm run dev -- @bpmn-io/form-js-viewer",
```
[the remark this answers is not in the document]
Comment (vsgoulart) [reply package.json:31]: it's fine for me, just wanted to understand it a bit better

File: packages/form-js-editor/src/features/properties-panel/Util.js [hunk 9]
```diff
@@ -56,6 +56,8 @@ export function textToLabel(text) {
   return null;
 }
 
+export const dotPathRegex = /^\w+(\.\w+)*$/;
```
Comment (vsgoulart) [inline packages/form-js-editor/src/features/properties-panel/Util.js:59]: I think this is slightly better, but it's up to you

```suggestion
export function isValidDotPath(path) {
  return /^\w+(\.\w+)*$/.test(path);
}
```

File: packages/form-js-viewer/assets/form-js-base.css [hunk 10]
```diff
@@ -312,6 +319,52 @@
   color: var(--color-text-lighter);
 }
 
+.fjs-container .fjs-form-field-group {
+  padding: 10px 6px 0 6px;
+  margin: 0 10px !important;
```
[the remark this answers is not in the document]
Comment (Skaiir) [reply packages/form-js-viewer/assets/form-js-base.css]: I've cleared a few !important keywords, and it does seem to have affected anything. Very nice 👀.

File: packages/form-js-viewer/src/core/PathRegistry.js [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+      if (child.claimCount === 0) {
+        node.children.splice(node.children.indexOf(child), 1);
+        break; // Abort early if claimCount reaches zero
+      }
+
+      node = child;
+    }
+  }
+
+  executeRecursivelyOnFields(field, fn, context = {}) {
```
Comment (pinussilvestrus) [inline packages/form-js-viewer/src/core/PathRegistry.js:91]: This one reads quite complicated, can we add some comments to give more context what's happening?

File: packages/form-js-viewer/src/core/PathRegistry.js [hunk 12]
```diff
@@ -0,0 +1,179 @@
+import { isArray } from 'min-dash';
+import { clone } from '../util';
+
+export default class PathRegistry {
```
Comment (pinussilvestrus) [inline packages/form-js-viewer/src/core/PathRegistry.js:4]: What can help a lot I guess is to add some documentation here about the whole purpose of the path registry module and the mental model behind it. This will help in some other places where the complexity of the nested data kicks in place (dragging, importer, ...)

File: packages/form-js-viewer/src/render/components/form-fields/Group.js [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        id={ prefixId(id, formId) }
+        label={ label } />
+      <Grid { ...fullProps } />
+    </div>
+  );
+}
+
+Group.config = {
+  type: 'group',
+  routed: true,
```
Comment (pinussilvestrus) [inline packages/form-js-viewer/src/render/components/form-fields/Group.js:32]: I had some struggles to understand what `routed` means, maybe `pathed` is a good alternative?

File: packages/form-js-viewer/test/spec/custom/index.js [hunk 14]
```diff
@@ -2,7 +2,7 @@ import { formFieldClasses } from '../../../src/render/components/Util';
 
 const type = 'button';
 
-function CustomButtonRenderer(props) {
```
Comment (pinussilvestrus) [inline packages/form-js-viewer/test/spec/custom/index.js:5]: Just a recommendation: The PR itself is already quite huge; maybe we can leave these unrelated changes out of it?

That does not mean that I don't appreciate the "refactor on the run" approach 🙂

**Review by pinussilvestrus — COMMENTED** [review]
Works super smooth! 👏 Implementation wise some things to improve, mostly to improve the understandability of this fairly complex implementation.

File: packages/form-js-editor/src/render/components/FormEditor.js [hunk 15]
```diff
@@ -383,15 +405,20 @@ export default function FormEditor(props) {
     eventBus.fire('formEditor.rendered');
   }, []);
 
-  const formRenderContext = {
+  const [ hoveredId, setHoveredId ] = useState(null);
+
+  const formRenderContext = useMemo(() => ({
```
[the remark this answers is not in the document]
Comment (Skaiir) [reply packages/form-js-editor/src/render/components/FormEditor.js:410]: Just good practice. Everything dependent on the formRenderContext would re-render every time it changes otherwise.

[the remark this answers is not in the document]
Comment (Skaiir) [reply packages/form-js-editor/src/render/components/FormEditor.js:410]: https://stackoverflow.com/a/62231233

File: packages/form-js-viewer/test/spec/render/components/form-fields/Button.spec.js [hunk 16]
```diff
@@ -81,7 +81,7 @@ describe('Button', function() {
     expect(config.type).to.eql('button');
     expect(config.label).to.eql('Button');
     expect(config.group).to.eql('action');
-    expect(config.keyed).to.be.true;
+    expect(config.keyed).to.be.false;
```
[the remark this answers is not in the document]
Comment (Skaiir) [reply packages/form-js-viewer/test/spec/render/components/form-fields/Button.spec.js:84]: I believe we used to put the action in the key a long time ago, because I've seen some old schemas that have it there.

File: packages/form-js-viewer/test/spec/custom/index.js (hunk 14 above)
Comment (Skaiir) [reply packages/form-js-viewer/test/spec/custom/index.js:5]: I fixed this because it was breaking our tests, due to button now not being keyed. Figured I might as well fix it completely :).

Comment (pinussilvestrus) [reply packages/form-js-viewer/test/spec/custom/index.js:5]: Fair enough! 👍

File: packages/form-js-carbon-styles/src/carbon-styles.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (pinussilvestrus) [reply packages/form-js-carbon-styles/src/carbon-styles.js:1166]: So far the carbon-styles are mainly to set the correct styles to be embedded into Tasklist, I'm not sure whether we should do the alignments over there. The reason is that Carbon is pretty strict on the alignments so I'd expect we will have the same alignment requirements in other Carbonized products as well.

What about not removing the margins on `.fjs-form-field-group` elements?

**Review by pinussilvestrus — APPROVED** [review]
Looks good to me 👏 

Let's [figure this one out](https://github.com/bpmn-io/form-js/pull/768/files#r1315679739) and get this out of the door 🚀

File: packages/form-js-carbon-styles/src/carbon-styles.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (vsgoulart) [reply packages/form-js-carbon-styles/src/carbon-styles.js:1166]: @pinussilvestrus @Skaiir Even though the original padding/margin comes from the Carbon grid, the external padding from the Carbon grid is an implementation detail from `form-js`
So right now I would have to always remove the padding because we have 3 different cases on Tasklist which 3 different paddings.

I would expect that when other teams, start using Carbon everywhere they might have different spacing. So I see no point in keeping the external margin if it will always be removed by everyone.

I think we should have no padding/margin on the outer containers. This way each user is free to set their own spacing in the `.fjs-container`

[the remark this answers is not in the document]
Comment (Skaiir) [reply packages/form-js-carbon-styles/src/carbon-styles.js:1166]: @vsgoulart But the thing is, with no padding, we cannot display the groups at all. Those padding are *required* at least within the self-contained application to show everything. 

You can see in the below graphics that the outlines are cut off by the theoretical zero-margin. So they aren't as much of an implementation detail as they are a requirement. 

grafik

**Review by vsgoulart — APPROVED** [review]
